applications/nrf_desktop/src/modules/ble_scan.c

```diff
@@ -132,29 +156,77 @@ static void scan_start(void)
 {
 	int err;
 
-	if (scanning) {
-		scan_stop();
+	struct bond_find_data bond_find_data = {
+		.conn_count = 0,
+		.peer_count = 0,
+	};
+
+	bt_foreach_bond(BT_ID_DEFAULT, bond_find,
+			&bond_find_data);
+	if (bond_find_data.conn_count == CONFIG_BT_MAX_CONN) {
+		LOG_INF("Max number of peers connected - scanning disabled");
+		return;
+	} else if (discovering_peer_conn) {
+		LOG_INF("Discovery in progress - scanning disabled");
+		return;
 	}
+
 	err = configure_filters();
 	if (err) {
 		LOG_ERR("Cannot set filters (err %d)", err);
 		goto error;
 	}
 
 	err = bt_scan_start(BT_SCAN_TYPE_SCAN_ACTIVE);
-	if (err) {
+	if (err == -EALREADY) {
+		LOG_WRN("Scanning already enabled");
+	} else if (err) {
 		LOG_ERR("Cannot start scanning (err %d)", err);
 		goto error;
+	} else {
+		LOG_INF("Scan started");
 	}
-	LOG_INF("Scan started");
-	scanning = true;
+
+	k_delayed_work_submit(&scan_stop_trigger, SCAN_DURATION_MS);
+	k_delayed_work_cancel(&scan_start_trigger);
+
 	return;
 
 error:
 	LOG_ERR("Scanning failed to start (err %d)", err);
 	module_set_state(MODULE_STATE_ERROR);
 }
 
+static void scan_start_trigger_fn(struct k_work *w)
+{
+	static_assert((SCAN_TRIG_TIMEOUT_MS > SCAN_TRIG_CHECK_MS) &&
+		      (SCAN_TRIG_CHECK_MS > 0), "");
+
+	scan_counter += SCAN_TRIG_CHECK_MS;
+	if (scan_counter > SCAN_TRIG_TIMEOUT_MS) {
+		scan_counter = 0;
+		scan_start();
+	} else {
+		k_delayed_work_submit(&scan_start_trigger, SCAN_TRIG_CHECK_MS);
+	}
+}
+
+static void scan_stop_trigger_fn(struct k_work *w)
+{
+	static_assert(SCAN_DURATION_MS > 0, "");
+
+	struct bond_find_data bond_find_data = {
```

Comment:
it seems we use this in multiple places now
maybe we should create header in utils and create a incline function there:
`size_t get_conn_count() {}`

Reply:
I will change it after we add reading VID and PID.
Let's leave it as it is now.

Reply:
discussed offline